### PR TITLE
mrc-3065 update particles states

### PR DESCRIPTION
**Pull request by hillalex**
Allow user to directly update particle states

### CODE REVIEW

File: inst/include/mode/solver.hpp [hunk 1]
```diff
@@ -106,6 +106,10 @@ class solver {
     t_ = t;
   }
 
+  void set_state(std::vector<double>::const_iterator state) {
+    stepper_.set_state(t_, state);
```
Comment (hillalex) [inline inst/include/mode/solver.hpp]: should we also reset the step size?

File: inst/include/mode/r/mode.hpp [hunk 2]
```diff
@@ -57,18 +57,59 @@ cpp11::sexp mode_solve(SEXP ptr, double end_time) {
...
 template <typename T>
-void mode_update_state(SEXP ptr, SEXP r_time) {
+void mode_update_state(SEXP ptr, SEXP r_state, SEXP r_time) {
   T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();
   if (r_time != R_NilValue) {
     auto time = validate_time(r_time);
     obj->set_time(time);
   }
+  if (r_state != R_NilValue) {
+    auto state = validate_state(r_state,
```
Comment (richfitz) [inline inst/include/mode/r/mode.hpp]: It's not obvious in the dust code, but what it tries to do is validate everything first, then set everything (where nothing is expected to throw) so that either all components of state get successfully set or none to, avoiding inconsistent state. You could do this here by doing each if the `if`'s twice, but the scope gets pretty nasty. Might be easiest to try and pick this issue up later though once all the state elements are in

File: inst/include/mode/solver.hpp (hunk 1 above)
Comment (richfitz) [reply inst/include/mode/solver.hpp]: I think that this needs to be an option that the user can control:

* If we reset time we _definitely_ should reset the step size
* If we reset the parameters (not yet implemented) we _probably_ should reset the step size
* If we reset the state we _possibly_ should reset the step size

The step size will adjust regardless (possibly through a set of rejected steps), but if the user has made only a small modification to the state we might not expect a large change in step size and the last step might be a better approximation of it.

Similar to the handling of "set initial", I think tidiest to add a `reset_step_size = NULL` option at the R-facing function, then if the user has not added an explicit argument, within your mode_update_state function resolve this based on the presentness of time/parameters/state.

File: inst/include/mode/mode.hpp [hunk 3]
```diff
@@ -55,6 +55,15 @@ class container {
...
+  void set_state(const std::vector<double>& state) {
+    const bool individual = state.size() == n_state() * n_particles_;
+    const size_t n = individual ? 1 : n_particles_;
+    auto it = state.begin();
+    for (size_t i = 0; i < n_particles_; ++i) {
+      solver_[i].set_state(it + (i / n) * n_state());
+    }
+  }
+
   void set_time(double time) {
```
Comment (richfitz) [inline inst/include/mode/mode.hpp]: Below, in set_time, we should not be setting `y` in the case where the user has provided state.  So this function probably needs a `set_initial_state` state argument; see the dust code for this, though our case is a bit nastier.

Probably we need to defer the time setting until the last part of the update function if it will control setting things like initial step sizes etc?

File: inst/include/mode/mode.hpp [hunk 4]
```diff
@@ -55,6 +55,15 @@ class container {
     }
   }
 
+  void set_state(const std::vector<double>& state) {
+    const bool individual = state.size() == n_state() * n_particles_;
+    const size_t n = individual ? 1 : n_particles_;
+    auto it = state.begin();
+    for (size_t i = 0; i < n_particles_; ++i) {
+      solver_[i].set_state(it + (i / n) * n_state());
```
Comment (richfitz) [inline inst/include/mode/mode.hpp]: A matter of taste, and quite probably compiles to the same assembly, but this might be a bit clearer:

```suggestion
      solver_[i].set_state(it);
      if (individual) {
        it += n_state();
      }
```

(if you go with this, you can drop `n`)

**Review by richfitz — CHANGES_REQUESTED** [review]

File: inst/include/mode/mode.hpp (hunk 3 above)
Comment (hillalex) [reply inst/include/mode/mode.hpp]: Ah yeah, of course, sort of forgot about this interaction. Ok, will try and re-jig this and add the `reset_step_size` argument as well

File: inst/include/mode/r/mode.hpp (hunk 2 above)
Comment (hillalex) [reply inst/include/mode/r/mode.hpp]: Oh that makes sense!

File: inst/include/mode/solver.hpp (hunk 1 above)
Comment (hillalex) [reply inst/include/mode/solver.hpp]: Ok, so if we reset *only* time, we should def reset step size, but if we reset time *and* state and the user hasn't specified, we should also default to resetting you think?

Comment (richfitz) [reply inst/include/mode/solver.hpp]: yes - following the most resetty of the different elements (so of time + anything else is set, and reset_step_size is NULL we reset the step size)

File: inst/include/mode/solver.hpp [hunk 5]
```diff
@@ -99,13 +99,38 @@ class solver {
     }
   }
 
-  void reset(double t, std::vector<double> y) {
-    stepper_.reset(t, y);
+  void reset(double t,
```
Comment (hillalex) [inline inst/include/mode/solver.hpp]: Could do away with this function as it's essentially the same as `set_state`, but would need to pass an iterator instead of a vector https://github.com/mrc-ide/mode/pull/8/files#diff-8c2c00c231ed8b1864f07acd80fcd795bec1b09543039b19c84bc3d6de9b3bf5R79 and https://github.com/mrc-ide/mode/pull/8/files#diff-6df4c22acbd9e0958ea70ddc605ef1d2b7369270ccea7732e6d3a24de77788a1R29

**Review by hillalex — COMMENTED** [review]
This has the desired behaviour now but is quite fiddly so any suggestions welcome!

File: inst/include/mode/solver.hpp (hunk 5 above)
Comment (richfitz) [reply inst/include/mode/solver.hpp]: GitHub is making those links hard to follow (I think they don't resolve unless my diff settings are the same as yours).  But if iterator vs vector is the only difference you could overload `set_state`:

```
  void set_state(double t,
                 const std::vector<double>& state,
                 bool reset_step_size) {
    set_state(t, state.begin(), reset_step_size);
  }
```

(retaining the existing definition for `solver::set_state`) and you can cut out the duplication

File: inst/include/mode/r/mode.hpp [hunk 6]
```diff
@@ -54,21 +54,95 @@ cpp11::sexp mode_solve(SEXP ptr, double end_time) {
...
+  }
+
+  bool reset_step_size = false;
+  if (r_reset_step_size == R_NilValue) {
+    reset_step_size = r_time != R_NilValue;
+  } else {
+    reset_step_size = cpp11::as_cpp<bool>(r_reset_step_size);
+  }
+
+  std::vector<double> time;
```
Comment (richfitz) [inline inst/include/mode/r/mode.hpp]: If you did:

```
auto time = validate_time(r_time);
```

and rewrote validate_time to do:

```
std::vector<double> validate_time(cpp11::sexp r_time) {
  if (r_time == R_NilValue) {
    return std::vector<double>(0);
  }
  cpp11::doubles time = cpp11::as_cpp<cpp11::doubles>(r_time);
  if (time.size() != 1) {
    cpp11::stop("Expected 'time' to be a scalar value");
  }
  std::vector<double> ret(1);
  std::copy_n(REAL(time.data()), 1, ret.begin());
  return ret;
}
```

and followed that pattern for all 4 inputs this function would clean up quite nicely I think.


```
template <typename T>
void mode_update_state(SEXP ptr, SEXP r_state, SEXP r_time,
                       SEXP r_set_initial_state,
                       SEXP r_reset_step_size) {
  T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();

  auto set_initial_state = validate_set_initial_state(r_set_initial_state, r_time, r_state);
  auto reset_step_size = validate_reset_step_size(r_reset_step_size, r_time);
  auto time = validate_time(r_time);
  auto state = validate_state(r_state);
  obj->update_state(time, state, set_initial_state, reset_step_size);
}
```

File: inst/include/mode/stepper.hpp [hunk 7]
```diff
@@ -114,13 +114,20 @@ class stepper {
     return std::sqrt(err / n);
   }
 
-  void reset(double t, std::vector<double> y0) {
+  void set_state(double t, std::vector<double> y0) {
     for (size_t i = 0; i < n; ++i) {
       y[i] = y0[i];
     }
     m.rhs(t, y, k1);
   }
```
Comment (richfitz) [inline inst/include/mode/stepper.hpp]:
```suggestion
  void set_state(double t, const std::vector<double>& y0) {
    set_state(t, y0.begin());
  }
```

File: inst/include/mode/mode.hpp [hunk 8]
```diff
@@ -55,10 +55,37 @@ class container {
...
+      const bool individual = state.size() == n_state() * n_particles_;
+      auto it = state.begin();
+      for (size_t i = 0; i < n_particles_; ++i) {
+        solver_[i].set_state(t, it, reset_step_size);
+        if (individual) {
+          it += n_state();
+        }
+      }
+    }
+    else if (set_initial_state) {
```
Comment (richfitz) [inline inst/include/mode/mode.hpp]: sorry, can you put both these else's on the same line as `}`?

File: inst/include/mode/mode.hpp [hunk 9]
```diff
@@ -55,10 +55,37 @@ class container {
...
+      auto y = m_.initial(t);
+      for (size_t i = 0; i < n_particles_; ++i) {
+        solver_[i].reset(t, y, reset_step_size);
+      }
+    }
+    else {
+      for (size_t i = 0; i < n_particles_; ++i) {
+        solver_[i].set_time(t);
+        if (reset_step_size) {
+          solver_[i].set_initial_step_size(t);
```
Comment (richfitz) [inline inst/include/mode/mode.hpp]: Should this method _not_ require `t`? Looks like it could just use `t_` or am I missing something there?

File: inst/include/mode/solver.hpp [hunk 10]
```diff
@@ -99,13 +99,38 @@ class solver {
...
+             std::vector<double> y,
+             bool reset_step_size) {
+    set_time(t);
+    stepper_.set_state(t, y);
+    if (reset_step_size) {
```
Comment (richfitz) [inline inst/include/mode/solver.hpp:111]: It feels like some of this logic is duplicated in mode.hpp; perhaps if we exposed the primative reset methods here (reset step size, reset statistics, set state, set time) and not support any compound reset here that may tidy the logic up?

**Review by richfitz — CHANGES_REQUESTED** [review]
Agree this is fiddly but that's also because there are many different valid ways of updating parts of the state - and we found in dust that different combinations turned up that we needed to support leading to a complicated method.

I've got a couple of suggestions here, in particular removing duplication by method overloading. Two methods can have the same name but differ in the argument types (they can't only differ in their return types) and that's fine.

**Review by hillalex — COMMENTED** [review]
Ok, I think this is a bit tidier now!

**Review by richfitz — APPROVED** [review]
